src/dp_transformers/dp_utils.py

```diff
@@ -174,14 +136,93 @@ class OpacusDPTrainer(Trainer):
         of sample-level DP, each sample can be represented by a unique author.
+        (iii) wrap the optimizer with Opacus' DPOptimizer/DistributedDPOptimizer
     """
-    def __init__(self, *args: Tuple, author_mapping: Optional[Sequence[Sequence[int]]]=None, **kwargs: Dict) -> None:
-        super().__init__(*args, **kwargs)
+    def __init__(
+        self,
+        model: Union[modeling_utils.PreTrainedModel, torch.nn.modules.module.Module] = None,
+        args: arguments.TrainingArguments = None,
+        train_dataset: Optional[torch.utils.data.dataset.Dataset] = None,
+        privacy_args: arguments.PrivacyArguments = None,
+        author_mapping: Optional[Sequence[Sequence[int]]] = None,
+        accountant_fn: callable = None,
+        use_prv_accountant: bool = True,
+        **kwargs: Dict
+    ) -> None:
+
+        self.train_args = args
+        self.privacy_args = privacy_args
+
         # Sample-level DP is equivalent to mapping each sample to a unique author. 
         if author_mapping is None:
-            author_mapping = [[i] for i in range(len(self.train_dataset))]
+            author_mapping = [[i] for i in range(len(train_dataset))]
         self.author_mapping = author_mapping
 
+        if not self.privacy_args.is_initialized:
+            self.privacy_args.initialize(
+                sampling_probability=self.sampling_probability,
+                num_steps=self.num_steps,
+                num_samples=len(self.author_mapping),
+            )
+
+        # Wrap model in DDP and GradSampleModule
+        if args.parallel_mode == training_args.ParallelMode.DISTRIBUTED:
+            logger.info(f"Wrapping the model with DPDDP in distributed training.")
+            model = opacus.distributed.DifferentiallyPrivateDistributedDataParallel(model)
+
+        model = GradSampleModule(model)
+
+        # Instantiate alternative accountant
+        if use_prv_accountant:
+            self.privacy_accountant = Accountant(
+                noise_multiplier=self.privacy_args.noise_multiplier,
+                sampling_probability=self.sampling_probability,
+                delta=self.privacy_args.target_delta,
+                eps_error=0.1,
+                max_compositions=self.num_steps
+            )
+            accountant_fn = lambda s: self.privacy_accountant.compute_epsilon(s)[2]
+        else:
+            accountant_fn = None
+
+        # Set up callback for accounting and handling grad acc
+        self.dp_callback = DPCallback(
+            noise_multiplier=self.privacy_args.noise_multiplier,
+            target_delta=self.privacy_args.target_delta,
+            sampling_probability=self.sampling_probability,
+            compute_epsilon=accountant_fn
+        )
+        super().__init__(model=model, args=args, train_dataset=train_dataset, callbacks=[self.dp_callback], **kwargs)
+
+        self.get_rdp_epsilon = lambda: self.dp_callback.accountant.get_epsilon(self.privacy_args.target_delta)  # RDP epsilon

```

Comment:
This seems a bit circular and not easy to follow. `Trainer.privacy_accountant` is a PRV accountant but `DPCallback.accountant` is a RDP accountant. Should we try to make the trainer have all the accountants and give them more interpretable names?